### PR TITLE
Add  set basic auth param

### PR DESCRIPTION
**Pull request by morarucostel**
Added the `-set-basic-auth` param to set the Basic Authorization header for upstreams.

## Motivation and Context
Problem and Context described in issue #394 

## How Has This Been Tested?

Tested in our environments.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: docs/configuration/configuration.md [hunk 1]
```diff
@@ -74,6 +74,7 @@ An example [oauth2_proxy.cfg]({{ site.gitweb }}/contrib/oauth2_proxy.cfg.example
 | `-pass-access-token` | bool | pass OAuth access_token to upstream via X-Forwarded-Access-Token header | false |
 | `-pass-authorization-header` | bool | pass OIDC IDToken to upstream via Authorization Bearer header | false |
 | `-pass-basic-auth` | bool | pass HTTP Basic Auth, X-Forwarded-User and X-Forwarded-Email information to upstream | true |
+| `-set-basic-auth` | bool | set HTTP Basic Auth information to upstream | true |
```
Comment (JoelSpeed) [inline docs/configuration/configuration.md]: This should be moved into the correct position for this list to be in alphabetical order
```suggestion
| `-set-basic-auth` | bool | set HTTP Basic Auth information in response (useful in Nginx auth_request mode) | true |
```

File: main.go [hunk 2]
```diff
@@ -41,6 +41,7 @@ func main() {
 	flagSet.Bool("set-xauthrequest", false, "set X-Auth-Request-User and X-Auth-Request-Email response headers (useful in Nginx auth_request mode)")
 	flagSet.Var(&upstreams, "upstream", "the http url(s) of the upstream endpoint, file:// paths for static files or static://<status_code> for static response. Routing is based on the path")
 	flagSet.Bool("pass-basic-auth", true, "pass HTTP Basic Auth, X-Forwarded-User and X-Forwarded-Email information to upstream")
+	flagSet.Bool("set-basic-auth", true, "set HTTP Basic Auth information to upstream")
```
Comment (JoelSpeed) [inline main.go]:
```suggestion
	flagSet.Bool("set-basic-auth", true, "set HTTP Basic Auth information in response (useful in Nginx auth_request mode)")
```

File: oauthproxy.go [hunk 3]
```diff
@@ -930,6 +932,14 @@ func (p *OAuthProxy) addHeadersForProxying(rw http.ResponseWriter, req *http.Req
 		}
 	}
 
+	if p.SetBasicAuth {
+		if session.User != "" {
+			rw.Header().Set("Authorization", fmt.Sprintf("Basic %s", b64.StdEncoding.EncodeToString([]byte(session.User+":"+p.BasicAuthPassword))))
```
Comment (JoelSpeed) [inline oauthproxy.go]: This option will clash with `SetAuthorizationHeader`, can we add some validation to ensure they aren't set simultaneously please?

File: options.go [hunk 4]
```diff
@@ -164,6 +165,7 @@ func NewOptions() *Options {
 		SetXAuthRequest:                  false,
 		SkipAuthPreflight:                false,
 		PassBasicAuth:                    true,
+		SetBasicAuth:                     true,
```
Comment (JoelSpeed) [inline options.go]: This should default to `false` otherwise this will always reflect peoples credentials back at them

**Comment by nickveenhof** [comment]
I'm a bit confused, this is a duplicate of mine. https://github.com/pusher/oauth2_proxy/pull/408. Can we maybe collaborate?

File: main.go (hunk 2 above)
Comment (morarucostel) [reply main.go]: fixed according to the feedback

File: oauthproxy.go (hunk 3 above)
Comment (morarucostel) [reply oauthproxy.go]: fixed according to the feedback

**Comment by morarucostel** [comment]
Pushed the changes according to the feedback.

**Comment by JoelSpeed** [comment]
@nickveenhof Does this PR solve the issue that you are trying to solve in #408? 
@ploxiln Since you were involved in the discussion in #413, would you mind casting an eye over this one too?

File: oauthproxy.go [hunk 5]
```diff
@@ -977,6 +979,12 @@ func (p *OAuthProxy) addHeadersForProxying(rw http.ResponseWriter, req *http.Req
 		} else {
 			rw.Header().Del("Authorization")
 		}
+	} else if p.SetBasicAuth {
+		if session.User != "" {
+			rw.Header().Set("Authorization", fmt.Sprintf("Basic %s", b64.StdEncoding.EncodeToString([]byte(session.User+":"+p.BasicAuthPassword))))
```
Comment (ploxiln) [inline oauthproxy.go]: This line is a bit long. To split it, I suggest something like:

```go
			auth_val := b64.StdEncoding.EncodeToString([]byte(session.User+":"+p.BasicAuthPassword))
			rw.Header().Set("Authorization", "Basic " + auth_val)
```

**Comment by ploxiln** [comment]
This follows the pattern of the existing `--set-authorization` option, so that seems like a fine justification.

I do think it is odd to use these headers, which are standard request headers, as response headers. If I was creating them for the first time, I might have used the header names `X-Auth-Bearer` and `X-Auth-Basic` or similar, to match the pattern of the even older `--set-xauthrequest` option which sets `X-Auth-Request-User` etc in the response. And if you combine `--set-xauthrequest` with another old flag `--pass-authorization-header`, then it sets `X-Auth-Request-Access-Token` in the response, so it _was_ possible to use the awkward nginx header vars logic to set `Authorization: Bearer <token>` after an `auth_request` even _without_ the `--set-authorization` option that was more recently added.

But, there has been no way to form the basic auth header in nginx, that I know of. No in-tree module provides a base64 function, and no way to set auth user/pass in the `ngx_http_proxy_module`. So given the not-unheard-of desire for this functionality, and the existing pattern of `--set-authorization`, I give my 👍 .

One other note: might it be confusing that `--set-authorization` takes precedence over `--set-basic-auth`? Maybe it should be the other way, because if you provide both flags, it seems like you want the Authorization response header, and you want it in Basic format.

**Comment by JoelSpeed** [comment]
@ploxiln Thanks for your review and insight, there's definitely some improvement (cleanup and deprecation) that could be done to this section of the code as you've suggested :) 

> One other note: might it be confusing that --set-authorization takes precedence over --set-basic-auth? Maybe it should be the other way, because if you provide both flags, it seems like you want the Authorization response header, and you want it in Basic format.

I would suggest the options become mutually exclusive via the options validation so we don't have  problem of precedence, WDYT?

**Comment by ploxiln** [comment]
> I would suggest the options become mutually exclusive via the options validation

Yes, that would make sense.

**Comment by morarucostel** [comment]
@JoelSpeed sorry it took longer for the changes, I was focusing on something else.
I believe I covered all the comments.

File: docs/configuration/configuration.md [hunk 6]
```diff
@@ -103,6 +103,7 @@ An example [oauth2-proxy.cfg]({{ site.gitweb }}/contrib/oauth2-proxy.cfg.example
 | `-session-store-type` | string | [Session data storage backend](configuration/sessions); redis or cookie | cookie |
 | `-set-xauthrequest` | bool | set X-Auth-Request-User, X-Auth-Request-Email and X-Auth-Request-Preferred-Username response headers (useful in Nginx auth_request mode) | false |
 | `-set-authorization-header` | bool | set Authorization Bearer response header (useful in Nginx auth_request mode) | false |
+| `-set-basic-auth` | bool | set HTTP Basic Auth information to in response (useful in Nginx auth_request mode) | true |
```
Comment (ploxiln) [inline docs/configuration/configuration.md:106]: "information to in response" - excess word "to"

File: CHANGELOG.md [hunk 7]
```diff
@@ -3,6 +3,7 @@
 ## Release Hightlights
 
 ## Important Notes
+- [#335] The session expiry for the OIDC provider is now taken from the Token Response (expires_in) rather than from the id_token (exp)
```
Comment (ploxiln) [inline CHANGELOG.md:6]: this appears unrelated to this pull-request, maybe there was some merge mixup

File: docs/configuration/configuration.md (hunk 6 above)
Comment (morarucostel) [reply docs/configuration/configuration.md]: Thanks, solved.

File: CHANGELOG.md (hunk 7 above)
Comment (morarucostel) [reply CHANGELOG.md]: Thanks. 
Yes, a mixup with a merge.

File: options.go [hunk 8]
```diff
@@ -243,6 +245,10 @@ func (o *Options) Validate() error {
 			"\n      use email-domain=* to authorize all email addresses")
 	}
 
+	if o.SetBasicAuth && o.SetAuthorization {
+		msgs = append(msgs, "mutually exclusive: set-basic-auth and set-authorization-header can not be set both on true")
```
Comment (JoelSpeed) [inline options.go:249]:
```suggestion
		msgs = append(msgs, "mutually exclusive: set-basic-auth and set-authorization-header can not both be true")
```

Comment (morarucostel) [reply options.go]: @JoelSpeed 
done and done.

File: CHANGELOG.md [hunk 9]
```diff
@@ -26,7 +26,7 @@
   - Binaries renamed from `oauth2_proxy` to `oauth2-proxy`
 - [#432](https://github.com/oauth2-proxy/oauth2-proxy/pull/432) Update ruby dependencies for documentation (@theobarberbany)
 - [#471](https://github.com/oauth2-proxy/oauth2-proxy/pull/471) Add logging in case of invalid redirects (@gargath)
-- [#462](https://github.com/oauth2-proxy/oauth2-proxy/pull/462) Allow HTML in banner message (@eritikass).
+- [#413](https://github.com/oauth2-proxy/oauth2-proxy/pull/413) Add -set-basic-auth param to set the Basic Authorization header for upstreams (@morarucostel).
```
Comment (JoelSpeed) [inline CHANGELOG.md:29]: Something went wrong here, need a new entry rather than overwriting the existing one

Comment (morarucostel) [reply CHANGELOG.md:30]: I don't know what happend, I don't recall overwriting an existing entry...
However I introduce re-introduced the previous one, so hopefully it fixed the CHANGELOG issue.
